plutus-tx/src/Language/PlutusTx/Ratio.hs

```diff
@@ -0,0 +1,205 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE TemplateHaskell     #-}
+{-# LANGUAGE NoImplicitPrelude    #-}
+{-# OPTIONS_GHC -fno-omit-interface-pragmas #-}
+{-# OPTIONS_GHC -fno-ignore-interface-pragmas #-}
+module Language.PlutusTx.Ratio(
+    Ratio(..)
+    , Rational
+    , (%)
+    , numerator
+    , denominator
+    , round
+    , truncate
+    , properFraction
+    , half
+    , fromGHC
+    , toGHC
+    -- * Misc.
+    , quotRem
+    , gcd
+    , signum
+    , abs
+    , reduce
+    ) where
+
+import qualified Language.PlutusTx.Applicative as P
+import qualified Language.PlutusTx.IsData      as P
+import qualified Language.PlutusTx.Data        as P
+import qualified Language.PlutusTx.Numeric     as P
+import qualified Language.PlutusTx.Eq          as P
+import qualified Language.PlutusTx.Lift        as P
+import qualified Language.PlutusTx.Ord         as P
+import qualified Language.PlutusTx.Bool        as P
+
+import           Language.PlutusTx.Functor
+import qualified Language.PlutusTx.Builtins as Builtins
+
+import qualified GHC.Real as Ratio
+import           Prelude (Bool(True), Integer, Maybe(..))
+
+data Ratio a = a :% a
+
+type Rational = Ratio Integer
+
+instance P.IsData a => P.IsData (Ratio a) where
+    {-# INLINABLE toData #-}
+    toData (n :% d) = P.Constr 0 [P.toData n, P.toData d]
+    {-# INLINABLE fromData #-}
+    fromData (P.Constr i [n, d]) | i P.== 0 = (:%) <$> P.fromData n P.<*> P.fromData d
+    fromData _ = Nothing
+
+instance P.Eq a => P.Eq (Ratio a) where
+    {-# INLINABLE (==) #-}
+    (n1 :% d1) == (n2 :% d2) = n1 P.== n2 P.&& d1 P.== d2
```

Comment:
`n1 P.* d2 == n2 P.* d1`?

Reply:
Yeah it assumes that the ratios are constructed using `reduce`. `GHC.Ratio.Ratio` actually uses the derived `Eq` instance: http://hackage.haskell.org/package/base-4.9.1.0/docs/src/GHC.Real.html#line-69